backend/forms/models/response.py

```diff
@@ -30,6 +33,102 @@ class Response(models.Model):
         verbose_name=_("IP address"),
     )
 
+    # denormalized fields
+    cached_dimensions = models.JSONField(
+        default=dict,
+        help_text="dimension slug -> list of value slugs",
+    )
+
+    # related fields
+    dimensions: models.QuerySet[ResponseDimensionValue]
+
+    @property
+    def survey(self) -> Survey | None:
+        return self.form.survey
+
+    def build_cached_dimensions(self) -> dict[str, list[str]]:
+        """
+        Used by ..handlers/dimension.py to populate cached_dimensions
+        """
+        # TODO should all survey dimensions always be present, or only those with values?
+        # TODO when dimensions are changed for an survey, refresh all cached_dimensions
+        survey = self.survey
+        if survey is None:
+            return {}
+
+        dimensions = {dimension.slug: [] for dimension in survey.dimensions.all()}
+        for sdv in self.dimensions.all():
+            dimensions[sdv.dimension.slug].append(sdv.value.slug)
```

Comment:
Could pull `self.dimensions.all()` to a local again, and also add `.select_related("dimension")`?